docs/async-await-proposal.md

```diff
@@ -0,0 +1,442 @@
+>   `EventLoopFuture` to obtain the `status` of the call.
+
+NIO types are not exposed. Asynchronous functions and properties are marked as
+`async` and the user makes use of the `await` keyword when using them.
+
+> 2. Code does not read in a straight line due to the need to provide a completion
+>   handler when making the call.
+
+While the above example is reasonably artificial, the response handling code can
+be placed after the code that is sending requests.
+
+#### Simple/safe wrappers
+
+The client API above provides maximum expressibility but has a notable drawback
+that was not present in the existing callback-based API. Specifically, in the
+server- and bidirectional-streaming cases, if the user does not consume the
+responses then waiting on the status will block indefinitely. This can be
+considered the converse of the drawback with the _existing_ server API that this
+proposal addresses.
+
+It is for this reason that the above proposed client APIs have slightly obscured
+names (e.g. `makeUpdateCall` instead of `update`) and we propose also generating
+additional, less expressive, but safer APIs. These APIs will not expose the RPC
+metadata (e.g. the status, initial metadata, trailing metadata), but will
+instead either simply return the response(s) or throw an error.
+
+In addition to avoiding the pitfall of the expressive counterparts, the client-
+and bidirectional-streaming calls provide the ability to pass an `AsyncSequence`
+of requests. In this way, the underlying library takes care of ensuring that no
+part of the RPC goes unterminated (both the request and response streams). It
+also offers an opportunity for users who have an `AsyncSequence` from which they
+are generating requests to make use of the combinators of `AsyncSequence` to not
+have to introduce unnecessary synchrony.
+
+We expect these will be sufficient for a lot of client use cases and, because
+they do not have the same pitfalls as their more expressive counterparts, we
+propose they be generated with the "plain names" of the RPC calls (e.g.
+`update`).
+
+For example, these are the additional APIs we propose to generate:
+
+```swift
+extension Echo_AsyncEchoClientProtocol {
+  public func get(
+    _ request: Echo_EchoRequest,
+    callOptions: CallOptions? = nil
+  ) async throws -> Echo_EchoResponse { ... }
+
+  public func collect<RequestStream>(
+    requests: RequestStream,
+    callOptions: CallOptions? = nil
+  ) async throws -> Echo_EchoResponse
+  where RequestStream: AsyncSequence, RequestStream.Element == Echo_EchoRequest { ... }
+
+  public func expand(
+    _ request: Echo_EchoRequest,
+    callOptions: CallOptions? = nil
+  ) -> GRPCAsyncStream<Echo_EchoResponse> { ... }
+
+  public func update<RequestStream>(
+    requests: RequestStream,
+    callOptions: CallOptions? = nil
+  ) -> GRPCAsyncStream<Echo_EchoResponse>
+  where RequestStream: AsyncSequence, RequestStream.Element == Echo_EchoRequest { ... }
```

Comment:
Signature looks good (that's the "akka grpc way" I linked/mentioned above, very happy to see it be offered as well here) 👍